Neurodamus-models to depend on neurodamus-py (bringing core)

ferdonline opened the pull request:
### Summary
As part of the repo reorganization (2021), neurodamus-core becomes officially an addendum to neurodamus-py

This PR changes the way neurodamus is built, and ultimately loaded, by setting neurodamus-py as a direct dependency of the models.

### Context
Neurodamus-core has became a submodule of neurodamus-py to support the long term goal of dropping it if favor of pure Python. Consequently it can/must no longer be considered a main package. 
We therefore make neurodamus-models to depend directly on neurodamus-py, since it brings it all, which streamlines the process of building, deploying and even loading. 
The end user will be able to simply load e.g. `neurodamus-neocortex` and have a fully functional environment with guaranteed cross-compatibility of the neocortex model, core and (Edit: Soon) neurodamus-py.
**Edit**: We disabled auto-loading neurodamus-py since it will bring the intel stack, which is not optimal because we may need to use other stacks (bglig-py...). These issues should be addresses in a dedicated ticket.

### Extra

Bumped neurodamus-py to 2.10.0, bringing
 - initial Neurodamulation support
 - Sonata config fixes
 - Some mandatory dependencies became optional, namely scipy is no longer required for normal workloads
 - etc

ferdonline committed:
Updating neurodamus recipes to depend on neurodamus-py

ferdonline committed:
Change recipe to take nd-core files from neurodamus-py

ferdonline committed:
Neurodamus all_deps default for bb5 deployments

pramodk commented on deploy/configs/packages.yaml:
```diff
@@ -74,6 +74,8 @@ packages:
       prefix: /usr
   petsc:
     variants: ~fortran
+  py-neurodamus:
+    variants: +all_deps
   python:
```
as Matthias pointed out, this should be moved to applications_hpc.yaml under packages key (see NSE example)

ferdonline committed:
Move spack configs to be deploy only

ferdonline committed:
Updated neurodamus recipes for backward compatibility

Keeping old function version_from_model_core_dep to enable building
latest stable deployed version depending still on neurodamus-core.

Changed recipes to call the function once per version for readability

matz-e commented on deploy/environments/applications_hpc.yaml:
```diff
@@ -14,8 +14,9 @@ spack:
         - neurodamus-mousify
         - parquet-converters
         - py-basalt
+        - py-libsonata
```
Why do we need a module for that?

ferdonline replied:
Oh, I abused this PR and added this ever-wanted module.

matz-e replied:
Then it should go into libraries, and also listed as a spec. After all, it's not HPC specific ;)

ferdonline committed:
Style

ferdonline committed:
Style on 3rd party recipe failing after vermin bump

ferdonline committed:
Move py-libsonata to libraries

matz-e commented on deploy/environments/libraries.yaml:
```diff
@@ -115,6 +116,7 @@ spack:
     - py-jinja2
     - py-jprops
     - py-lazy-property
+    - py-libsonata%intel
```
I don't think we should have a generic `%intel` module around when *all* other modules we provide a `%gcc`. That just spells like trouble somewhere down the road.

ferdonline replied:
In simulation we mostly use the intel version, but I see your point, we probably want the gcc one when user loads manually. Changing then

ferdonline committed:
py-libsonata module %gcc

ferdonline committed:
Bump neurodamus to 2.9.0

jorblancoa commented on var/spack/repos/builtin/packages/neurodamus-mousify/package.py:
```diff
@@ -16,15 +19,11 @@ class NeurodamusMousify(NeurodamusModel):
-        ('1.1', '3.0.2'),
-    )
-    version_from_model_core_deps(model_core_dep_v)
+    version_from_model_ndpy_dep('1.6', '2.9.0a')
```
This needs to be 2.9.0 now right?

jorblancoa commented on var/spack/repos/builtin/packages/neurodamus-neocortex/package.py:
```diff
@@ -14,29 +17,17 @@ class NeurodamusNeocortex(NeurodamusModel):
+    version('develop', branch='main', submodules=True, get_full_repo=True)
+    # IMPORTANT: Register new versions only using version_from_model_*
+    # Final version name is combined e.g. "1.0-3.0.1"
+    version_from_model_ndpy_dep('1.6', '2.9.0a')
```
same, 2.9.0

jorblancoa commented on var/spack/repos/builtin/packages/neurodamus-hippocampus/package.py:
```diff
@@ -16,21 +19,8 @@ class NeurodamusHippocampus(NeurodamusModel):
-    version('0.3', tag='0.3', submodules=True, get_full_repo=False)
-    version('0.2', tag='0.2', submodules=True, get_full_repo=False)
-    version('0.1', tag='0.1', submodules=True, get_full_repo=False)
+    version_from_model_ndpy_dep('1.6', '2.9.0a')
```
py-neurodamus version should be 2.9.0 (see others)

ferdonline committed:
Use latest neurpdamuspy_v and make it globally configurable

jorblancoa commented on var/spack/repos/builtin/packages/neurodamus-model/package.py:
```diff
@@ -13,25 +13,36 @@
 _CORENRN_MODLIST_FNAME = "coreneuron_modlist.txt"
 _BUILD_NEURODAMUS_FNAME = "build_neurodamus.sh"
 
+PYNEURODAMUS_DEFAULT_V = '2.9.0'
```
Maybe we should do the same for NEURODAMUSCORE_DEFAULT_V? I guess we would need it for a bit

ferdonline replied:
I don't think we need that anymore, I only left one previous version depending on core

jorblancoa reviewed (dismissed):
LGTM!

ferdonline committed:
Merge branch 'develop' into neurodamus/new_recipes

Conflicts:
	bluebrain/repo-bluebrain/packages/neurodamus-hippocampus/package.py
	bluebrain/repo-bluebrain/packages/neurodamus-mousify/package.py
	bluebrain/repo-bluebrain/packages/neurodamus-neocortex/package.py
	bluebrain/repo-bluebrain/packages/neurodamus-thalamus/package.py
	deploy/environments/libraries.yaml
	var/spack/repos/builtin/packages/py-charm4py/package.py

ferdonline committed:
Try with older scipy

ferdonline committed:
Use neurodamus-py%gcc

ferdonline committed:
Neurodamus to depend on previous scipy, compt with intel compiler

ferdonline committed:
Attempt to fix module issues

ferdonline committed:
Dont run dep neurodamus-model and py-neurodamus yet

matz-e commented on bluebrain/deployment/environments/applications_hpc.yaml:
```diff
@@ -14,8 +14,9 @@ spack:
         - neurodamus-mousify
         - parquet-converters
         - py-basalt
+        - 'py-libsonata%gcc'
```
This is still not HPC, really. Since there isn't a spec associated with it, it should be removed from the module whitelist (module whitelist is a strict subset of specs installed).

ferdonline replied:
~~Why doesnt the `py-libsonata` spec below qualify?~~ Forgot to push it.
Anyway true that it isn't really hpc, so probably a better fit for `application_libraries`

ferdonline committed:
Bump neurodamus for 2.10, fixing deps

matz-e commented on bluebrain/repo-bluebrain/packages/neurodamus-model/package.py:
```diff
@@ -105,7 +123,7 @@ def merge_hoc_mod(self, spec, prefix):
         so that incremental builds can actually happen.
         """
-        core_prefix = spec['neurodamus-core'].prefix
+        core_prefix = spec['py-neurodamus'].prefix
```
How does this work with things using with `version_from_model_core_dep`? Shouldn't `spec['py-neurodamus']` then just be `None`?

This bit seems the only problematic thing to me. IMO it would be cleaner to have PyDamus and HocDamus both providing a "virtual" NeuroDamus, and then use this virtual dependency to access stuff in the spec.

Of course, switching between the different providers would require a `^py-neurodamus` etc in the environment.

ferdonline replied:
Hum... broken backward compat. Your idea looks fancier but IMO we can get away with a simpler check for what dependency is available. Let me draft smtg for that

matz-e commented on bluebrain/repo-bluebrain/packages/py-neurodamus/package.py:
```diff
@@ -29,21 +31,25 @@ class PyNeurodamus(PythonPackage):
     depends_on('py-libsonata',     type='run', when='@2.5.3:')
     depends_on('py-morphio',       type='run', when='@2.6.0:')
-    depends_on('py-scipy',         type='run', when='+all_deps@2.5.3:')
+    depends_on('py-scipy@1.5.4',   type='run', when='+all_deps@2.5.3:')
```
Do we still need that version pin?

ferdonline replied:
Nope. Actually this dependency could go away and loaded on demand for LFP and such

ferdonline committed:
Cleanup. Ndamus building well with ^nd-core

ferdonline committed:
Bump neurodamus-py dep to 2.10

ferdonline committed:
import order!

matz-e reviewed (approved):
Looks reasonable. I'd prefer if the structure were simpler still, but we can leave that for future work.
